From f82034f0ac7f3a89b360fbf30bdd61e94aaee8fa Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@gmail.com>
Date: Thu, 22 May 2025 14:54:40 +0200
Subject: [PATCH] don't leak lcms tone curves

cmsStageAllocToneCurves duplicates them, it does not take ownership


(cherry picked from commit 14604f68f835c88427fcef0a9e7380d5a182a182)

Co-authored-by: Xaver Hugl <xaver.hugl@kde.org>
---
 src/colors/colordevice.cpp       |  2 --
 src/core/colortransformation.cpp |  5 +++++
 src/core/iccprofile.cpp          | 11 ++++-------
 3 files changed, 9 insertions(+), 9 deletions(-)

diff --git a/src/colors/colordevice.cpp b/src/colors/colordevice.cpp
index ff2de5713fc..c58f2bae1a9 100644
--- a/src/colors/colordevice.cpp
+++ b/src/colors/colordevice.cpp
@@ -14,8 +14,6 @@
 
 #include <QTimer>
 
-#include <lcms2.h>
-
 namespace KWin
 {
 
diff --git a/src/core/colortransformation.cpp b/src/core/colortransformation.cpp
index c6d063e05b9..03a688f4003 100644
--- a/src/core/colortransformation.cpp
+++ b/src/core/colortransformation.cpp
@@ -86,6 +86,11 @@ std::unique_ptr<ColorTransformation> ColorTransformation::createScalingTransform
     auto g = cmsBuildParametricToneCurve(nullptr, 2, curveParams.data());
     curveParams = {1.0, scale.z(), 0.0};
     auto b = cmsBuildParametricToneCurve(nullptr, 2, curveParams.data());
+    const auto guard = qScopeGuard([r, g, b]() {
+        cmsFreeToneCurve(r);
+        cmsFreeToneCurve(g);
+        cmsFreeToneCurve(b);
+    });
     if (!r || !g || !b) {
         qCWarning(KWIN_CORE) << "Failed to build tone curves";
         return nullptr;
diff --git a/src/core/iccprofile.cpp b/src/core/iccprofile.cpp
index 17be36c93d5..4707224ef67 100644
--- a/src/core/iccprofile.cpp
+++ b/src/core/iccprofile.cpp
@@ -267,14 +267,8 @@ IccProfile::Expected IccProfile::load(const QString &path)
     std::shared_ptr<ColorTransformation> vcgt;
     cmsToneCurve **vcgtTag = static_cast<cmsToneCurve **>(cmsReadTag(handle, cmsSigVcgtTag));
     if (vcgtTag && vcgtTag[0]) {
-        // Need to duplicate the VCGT tone curves as they are owned by the profile.
-        cmsToneCurve *toneCurves[] = {
-            cmsDupToneCurve(vcgtTag[0]),
-            cmsDupToneCurve(vcgtTag[1]),
-            cmsDupToneCurve(vcgtTag[2]),
-        };
         std::vector<std::unique_ptr<ColorPipelineStage>> stages;
-        stages.push_back(std::make_unique<ColorPipelineStage>(cmsStageAllocToneCurves(nullptr, 3, toneCurves)));
+        stages.push_back(std::make_unique<ColorPipelineStage>(cmsStageAllocToneCurves(nullptr, 3, vcgtTag)));
         vcgt = std::make_shared<ColorTransformation>(std::move(stages));
     }
 
@@ -403,6 +397,9 @@ IccProfile::Expected IccProfile::load(const QString &path)
     }
     std::vector<std::unique_ptr<ColorPipelineStage>> stages;
     stages.push_back(std::make_unique<ColorPipelineStage>(cmsStageAllocToneCurves(nullptr, toneCurves.size(), toneCurves.data())));
+    for (auto toneCurve : toneCurves) {
+        cmsFreeToneCurve(toneCurve);
+    }
     const auto inverseEOTF = std::make_shared<ColorTransformation>(std::move(stages));
     return std::make_unique<IccProfile>(handle, Colorimetry(red, green, blue, white), std::move(bToA0), std::move(bToA1), inverseEOTF, vcgt, relativeBlackPoint, maxBrightness);
 }
-- 
GitLab

